src/handlers/cors.js

```diff
@@ -1,17 +1,79 @@
+    allowCredentials = true,
+    allowedHeaders = ['Content-Type'],
+    allowedExposeHeaders = ['WWW-Authenticate', 'Server-Authorization'],
+    maxAge = '600',
+    optionsSuccessStatus = 204,
+    terminatePreflight = false
+  }) {
   return async (ctx, next) => {
-    const { origin } = ctx.request.headers;
-
-    if (origin) {
-      if (allowedOrigins[0] === '*' || allowedOrigins.indexOf(origin) !== -1) {
-        ctx.set('Access-Control-Allow-Credentials', 'true');
-        ctx.set('Access-Control-Allow-Origin', origin);
-        ctx.set('Access-Control-Allow-Methods', 'GET,PUT,POST,PATCH,DELETE,HEAD,OPTIONS');
-        ctx.set('Access-Control-Allow-Headers', 'Content-Type');
-        ctx.set('Access-Control-Expose-Headers', 'WWW-Authenticate,Server-Authorization');
+    const method = ctx.request.method
+    const origin = ctx.request.headers.origin
+    const requestHeaders = ctx.request.headers['access-control-request-headers']
+
+
+    if (method === 'OPTIONS') {
+      configureOrigin(ctx, origin, allowedOrigins);
+      configureCredentials(ctx, allowCredentials);
+      configureMethods(ctx, allowedMethods);
+      configureAllowedHeaders(ctx, requestHeaders, allowedHeaders);
+      configureMaxAge(ctx, maxAge);
+      configureExposedHeaders(ctx, allowedExposeHeaders);
+      if (terminatePreflight) {
+        ctx.status = optionsSuccessStatus;
+        ctx.set('Content-Length', '0');
+        ctx.body = '';
+        return;
       }
+    } else {
+      configureOrigin(ctx, origin, allowedOrigins);
+      configureCredentials(ctx, allowCredentials);
+      configureExposedHeaders(ctx, allowedExposeHeaders);
     }
-
     await next(ctx);
   };
 };
+
+
+function configureOrigin (ctx, origin, allowedOrigins) {
+  if (allowedOrigins === '*' || allowedOrigins.length && allowedOrigins[0] === '*') {
+    ctx.set('Access-Control-Allow-Origin', '*');
+  } else if (typeof allowedOrigins === 'string' && origin === allowedOrigins) {
+    ctx.set('Access-Control-Allow-Origin', origin);
+    ctx.set('Vary', 'Origin');
+  } else if (Array.isArray(allowedOrigins) && allowedOrigins.indexOf(origin) !== -1) {
+    ctx.set('Access-Control-Allow-Origin', origin);
+    ctx.set('Vary', 'Origin');
+  }
+}
+
+function configureCredentials (ctx, allowCredentials) {
+  ctx.set('Access-Control-Allow-Credentials', allowCredentials?'true':'false');
+}
+
+function configureMethods (ctx, allowedMethods) {
+  ctx.set('Access-Control-Allow-Methods', allowedMethods.join(','));
+}
+
+function configureAllowedHeaders (ctx, requestHeaders, allowedHeaders) {
+  if (allowedHeaders.length === 0 && requestHeaders) {
+    ctx.set('Access-Control-Allow-Headers', requestHeaders) //allowedHeaders wasn't specified, so reflect the request headers
+  } else if (allowedHeaders.length) {
+    ctx.set('Access-Control-Allow-Headers', allowedHeaders.join(','))
+  }
+}
+
+function configureMaxAge (ctx, maxAge) {
+  if (maxAge) {
+    ctx.set('Access-Control-Max-Age', ""+maxAge)
```

Comment:
Think the eslint-rules won't like the string concatenation. Not sure if ctx.set will stringify the value or not? I'll have to check how koa does it..

Reply:
Looks like koa automatically stringifies the headers before they are sent, so it "should" work the same way here..

Reply:
Ok, I will remove the quotes.